src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,692 @@
+            port_opacity.set_target_value(1.0);
+            port_size.set_target_value(HIGHLIGHT_SIZE);
+        });
+
+        eval_ show_normal ([port_opacity,port_size] {
+            port_opacity.set_target_value(0.5);
+            port_size.set_target_value(BASE_SIZE);
+        });
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+/// Frp API of the `OutPutPorts`.
+#[derive(Clone,CloneRef,Debug)]
+pub struct Frp {
+    /// Update the size of the `OutPutPorts`. Should match the size of the parent node for visual
+    /// correctness.
+    pub set_size        : frp::Source<Vector2<f32>>,
+    /// Emitted whenever one of the ports receives a `MouseDown` event. The `PortId` indicates the
+    /// source port.
+    pub port_mouse_down : frp::Stream<PortId>,
+    /// Emitted whenever one of the ports receives a `MouseOver` event. The `PortId` indicates the
+    /// source port.
+    pub port_mouse_over : frp::Stream<PortId>,
+    /// Emitted whenever one of the ports receives a `MouseOut` event. The `PortId` indicates the
+    /// source port.
+    pub port_mouse_out  : frp::Stream<PortId>,
+
+    on_port_mouse_down  : frp::Source<PortId>,
+    on_port_mouse_over  : frp::Source<PortId>,
+    on_port_mouse_out   : frp::Source<PortId>,
+}
+
+impl Frp {
+    fn new(network: &frp::Network) -> Self {
+        frp::extend! { network
+            def set_size           = source();
+            def on_port_mouse_down = source();
+            def on_port_mouse_over = source();
+            def on_port_mouse_out  = source();
+
+            let port_mouse_down = on_port_mouse_down.clone_ref().into();
+            let port_mouse_over = on_port_mouse_over.clone_ref().into();
+            let port_mouse_out  = on_port_mouse_out.clone_ref().into();
+        }
+        Self{set_size,on_port_mouse_down,on_port_mouse_over,on_port_mouse_out,
+             port_mouse_down,port_mouse_over,port_mouse_out}
+    }
+}
+
+
+
+// =======================
+// === OutputPortsData ===
+// =======================
+
+/// Internal data of the `OutputPorts`.
+#[derive(Debug)]
+pub struct OutputPortsData {
+    display_object : display::object::Instance,
+    logger         : Logger,
+    size           : Cell<Vector2<f32>>,
+    gap_width      : Cell<f32>,
+    ports          : RefCell<ShapeView>,
+}
+
+impl OutputPortsData {
+
+    fn new(scene:&Scene, number_of_ports:NonZeroU32) -> Self {
+        let logger         = Logger::new("OutputPorts");
+        let display_object = display::object::Instance::new(&logger);
+        let size           = Cell::new(Vector2::zero());
+        let gap_width      = Cell::new(SEGMENT_GAP_WIDTH);
+        let ports          = ShapeView::new(number_of_ports,&logger,scene);
+        let ports          = RefCell::new(ports);
+
+        OutputPortsData {display_object,logger,size,ports,gap_width}.init()
+    }
+
+    fn init(self) -> Self {
+        self.ports.borrow().display_object().set_parent(&self.display_object);
```

Comment:
do not use `display_obeject()` explicitelly. just skip it

`self.ports.borrow().set_parent(&self);`